### PR TITLE
WebGL RequestCountlyHelper Bug Fix

### PR DESCRIPTION
**Pull request by AliRKat**
Bug: in RequestCountlyHelper.cs at line 270 there is `StartProcessRequestRoutine` which is used for WebGL builds. Issue was if the line 280 
```
_monoBehaviour.StartCoroutine(ProcessRequestCoroutine(uri, data, (response) => {
            try {
                tcs.SetResult(response);
                // 'IsObjectMonoBehaviour can only be called from the main thread'
            }));
```
was called outside of the main thread, it was giving an error of 'IsObjectMonoBehaviour can only be called from the main thread' which was causing the `ProcessQueue` method to stuck since `_isQueueBeingProcess` boolean was never turning to true again so SDK was not able to send anything afterwards.

Fix uses the `CountlyMainThreadHandler` to move the task to the main thread. So it doesn't cause that error and return a response so `ProcessQueue` method can also move on and update the flag afterwards.
Also added checks and added/updated logs to identify any other possible issue.

The bug was reproducible within the example app. After the fix within the example app, it didn't occur too.

Tests can be added too but I would love some guidance on what way to test this is the best

There are also issues with Session Service, which is moved to another PR

### CODE REVIEW

**Comment by AliRKat** [comment]
[[Unity] WebGL Request bug fix](https://trello.com/c/uI7l8Rtb/1919-unity-webgl-request-bug-fix)

**Review by arifBurakDemiray — CHANGES_REQUESTED** [review]
So Ali-dono
lets move the coroutine impl for session problem to another PR to cook this more 
Currently PR is a little complex to understand because the subject is complex.
Lets divide and conquer it ⚔️

**Comment by AliRKat** [comment]
> So Ali-dono lets move the coroutine impl for session problem to another PR to cook this more Currently PR is a little complex to understand because the subject is complex. Lets divide and conquer it ⚔️

@arifBurakDemiray thank you, i was also thinking about separating those but couldn't be sure, would love @turtledreams opinion too

**Review by arifBurakDemiray — APPROVED** [review]

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs [hunk 1]
```diff
@@ -75,29 +73,45 @@ internal void AddRequestToQueue(CountlyRequestModel request)
...
                 return;
             }
 
             _isQueueBeingProcess = true;
-            CountlyRequestModel[] requests = _requestRepo.Models.ToArray();
+            CountlyRequestModel[] requests = null;
+
+            try {
+                requests = _requestRepo.Models.ToArray();
+                Log.Verbose("[RequestCountlyHelper] ProcessQueue, requests: " + requests.Length);
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:85]: request count*

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs [hunk 2]
```diff
@@ -75,29 +73,45 @@ internal void AddRequestToQueue(CountlyRequestModel request)
...
+            } catch (Exception ex) {
+                Log.Warning($"[RequestCountlyHelper] ProcessQueue, exception occurred while converting Models to array. Exception: {ex}");
+                _isQueueBeingProcess = false;
+                return;
+            }
 
-            Log.Verbose("[RequestCountlyHelper] Process queue, requests: " + requests.Length);
+            // make sure that flag is corrected and exit
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:92]: maybe another verbose log here

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs [hunk 3]
```diff
@@ -109,18 +123,21 @@ private async Task<CountlyResponse> ProcessRequest(CountlyRequestModel model)
             bool shouldPost = _config.IsForcedHttpPostEnabled() || model.RequestData.Length > 2000;
 
 #if UNITY_WEBGL
-            // There is not HTTP GET for webGL. We always do a HTTP POST
+            // There is not HTTP GET for WebGL. We always do a HTTP POST
+            Log.Debug($"[RequestCountlyHelper] ProcessRequest, Calling StartProcessRequestRoutine for WebGL");
             return await StartProcessRequestRoutine(_countlyUtils.ServerInputUrl, model.RequestData);
 #else
             if (shouldPost) {
+                Log.Debug($"[RequestCountlyHelper] ProcessRequest, Using HTTP POST");
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:131]: cant see the shouldPost is legit or not, double check pls

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs [hunk 4]
```diff
@@ -153,7 +170,7 @@ private string AddChecksum(string query)
         /// <returns></returns>
         internal async Task<CountlyResponse> GetAsync(string uri, string data)
         {
-            Log.Verbose("[RequestCountlyHelper] GetAsync request: " + uri + " params: " + data);
+            Log.Verbose("[RequestCountlyHelper] GetAsync, calling with request: " + uri + " params: " + data);
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:173]: cmon each and every one should have [ ]

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs [hunk 5]
```diff
@@ -308,12 +339,11 @@ private bool IsSuccess(CountlyResponse countlyResponse)
                     if (json.ContainsKey("result")) {
                         return true;
                     }
-                } catch(JsonException) {
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:311]: log also if req has wrong status code

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs (hunk 3 above)
Comment (AliRKat) [reply Assets/Countly/Plugins/CountlySDK/Helpers/RequestCountlyHelper.cs:131]: `bool shouldPost = _config.IsForcedHttpPostEnabled() || model.RequestData.Length > 2000;`
lgtm. also it was working fine, what concerned you?

**Pull request merged**